src/openfermion/ops/_code_operator.py

```diff
@@ -0,0 +1,298 @@
+    """
+
+    def __init__(self, encoding, decoding):
+        """ Initialization of a binary code.
+
+        Args: 
+            encoding (array or list): nested lists or binary 2D-array
+            decoding (array or list): list of SymbolicBinary, list-like or str
+
+        Raises: 
+            TypeError: non-list, array like encoding or decoding, unsuitable
+            symbolicBinary generators,
+            BinaryCodeError: in case of decoder/encoder size mismatch or decoder
+            size, qubits indexed mismatch
+        """
+
+        if not isinstance(encoding, (numpy.ndarray, list, tuple)):
+            raise TypeError('encoding must be a list, array or tuple .')
+
+        if not isinstance(decoding, (numpy.ndarray, list, tuple)):
+            raise TypeError('decoding must be a list, array or tuple .')
+
+        self.enc = scipy.sparse.csc_matrix(encoding)
+        self.qubits, self.orbitals = numpy.shape(encoding)
+
+        if (self.orbitals != len(decoding)):
+            raise BinaryCodeError(
+                'size mismatch, decoder and encoder should have the same'
+                ' first dimension')
+
+        decoder_qubits = set()
+        self.dec = []
+
+        for symbolic_binary in numpy.array(decoding):
+
+            if isinstance(symbolic_binary, (tuple, list, str)):
+                symbolic_binary = SymbolicBinary(symbolic_binary)
+            if isinstance(symbolic_binary, SymbolicBinary):
+                self.dec.append(symbolic_binary)
+                decoder_qubits = decoder_qubits | set(
+                    symbolic_binary.enumerate_qubits())
+            else:
+                raise TypeError(
+                    'decoder component provided '
+                    'is not a suitable for SymbolicBinary',
+                    symbolic_binary)
+
+        if len(decoder_qubits) != self.qubits:
+            raise BinaryCodeError(
+                'decoder and encoder provided has different number of qubits')
+
+        if max(decoder_qubits) + 1 > self.qubits:
+            raise BinaryCodeError('decoder is not indexing some qubits. Qubits'
+                                  'indexed are: {}'.format(decoder_qubits))
+
+        self.dec = numpy.array(self.dec)
+
+    def __iadd__(self, appendix):
+        """ In-place appending a binary code with +=.
+        
+        Args:
+            appendix (BinaryCode): The code to append to the present one. 
+
+        Returns (BinaryCode): a global binary code with size n1+n2, N1+N2
+
+        Raises:
+            TypeError: appendix must be a BinaryCode
+        """
+        if not isinstance(appendix, BinaryCode):
+            raise TypeError('argument must be a BinaryCode.')
+
+        self.dec = numpy.append(self.dec,
+                                _shift_decoder(appendix.dec, self.qubits))
+        self.enc = scipy.sparse.bmat([[self.enc, None], [None, appendix.enc]])
+        self.qubits, self.orbitals = numpy.shape(self.enc)
```

Comment:
In other parts of the code we call these things "n_orbitals" and "n_qubits". You might also consider saying "modes" instead of "orbitals" because if this is used with other parts of OpenFermion there might be an ambiguity versus spatial orbitals and spin-orbitals (usually there are twice as many spin-orbitals than orbitals) but we want this code to be used for models both with and without spin degrees of freedom.